File: leetcode/cpp/merge_intervals.cpp
// Copyright 2018 Leo - All Rights Reserved
// Author: Leo
// Best solution time: O(nlogn)
// Best solution space: O(1)

// Solution dependencies
#include<cmath>
#include<string>
#include<vector>

using std::max;
using std::string;
using std::vector;

struct Interval {
  int start;
  int end;
  Interval() : start(0), end(0) {}
  Interval(int s, int e) : start(s), end(e) {}
};

class Solution {
public:
  vector<Interval> merge(vector<Interval>& intervals) {
    if (intervals.empty()) return vector<Interval>();

    // inplace sort by start
    sort(intervals.begin(), intervals.end(),
         [](const Interval &a, const Interval &b) -> bool {
           return a.start < b.start;
         });

    // inserting into new array
    vector<Interval> merged_intervals = {intervals[0]};
    for (const auto &interval : intervals) {
      int &last_interval_end = merged_intervals.back().end;
      if (interval.start <= last_interval_end) {
        last_interval_end = max(last_interval_end, interval.end);
      } else {
        merged_intervals.emplace_back(interval);
      }
    }
    return merged_intervals;
  }
};

File: leetcode/java/MergeIntervals.java
import java.util.ArrayList;
import java.util.List;

public class MergeIntervals {
  public List<Interval> merge(List<Interval> intervals) {
    if (intervals.size() <= 1) {
      return intervals;
    }

    intervals.sort((i1, i2) -> i1.start == i2.start ? i1.end - i2.end : i1.start - i2.start);

    List<Interval> output = new ArrayList<>();
    int start = intervals.get(0).start;
    int end = intervals.get(0).end;
    for (int i = 1; i < intervals.size(); i++) {
      Interval cur = intervals.get(i);
      if (cur.start > end) {
        output.add(new Interval(start, end));
        start = cur.start;
        end = cur.end;
      } else {
        end = Math.max(end, cur.end);
      }
    }
    output.add(new Interval(start, end));

    return output;
  }
}
